### PR TITLE
Fixed abandon faction camp error

### LINKED ISSUE #45987: Escaping from naming a faction camp results in a broken one that can't be abandoned
# Describe the bug

I asked my friendly NPC to create a faction camp and pressed escape when i was prompted to name it. It was created regardless. Now when i try to abandon it, it shows this and doesn't proceed:
```
18:49:35.814 ERROR : (error message will follow backtrace)
    /usr/bin/cataclysm(_Z21debug_write_backtraceRSo+0x3a) [0x55963e1a100a]
    /usr/bin/cataclysm(_Z8DebugLog10DebugLevel10DebugClass+0x2d9) [0x55963e1a2d69]
    /usr/bin/cataclysm(_Z12realDebugmsgPKcS0_S0_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x67) [0x55963e1a6a37]
    /usr/bin/cataclysm(_Z12realDebugmsgIJRKiS1_S1_EEvPKcS3_S3_S3_DpOT_+0x355) [0x55963e4bbae5]
    /usr/bin/cataclysm(_ZNK3map13get_submap_atERK8tripoint+0xe1) [0x55963e83e9e1]
    /usr/bin/cataclysm(_ZN3map18remove_submap_campERK8tripoint+0x12) [0x55963e8472f2]
    /usr/bin/cataclysm(_ZN8basecamp12abandon_campEv+0x4c2) [0x55963e2cba62]
    /usr/bin/cataclysm(_ZN13talk_function12abandon_campER3npc+0x5a) [0x55963ec8d26a]
    /usr/bin/cataclysm(_ZNK13talk_effect_t5applyER8dialogue+0x70) [0x55963ec56700]
    /usr/bin/cataclysm(_ZN8dialogue3optER15dialogue_windowRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERK10talk_topic+0x30a4) [0x55963ec728c4]
    /usr/bin/cataclysm(_ZN6avatar7talk_toESt10unique_ptrI6talkerSt14default_deleteIS1_EEbb+0x3e6) [0x55963ec734e6]
    /usr/bin/cataclysm(_ZN4game4chatEv+0x1c96) [0x55963ec75af6]
    /usr/bin/cataclysm(_ZN4game13handle_actionEv+0x1ab3) [0x55963e43f003]
    /usr/bin/cataclysm(_ZN4game7do_turnEv+0x74a) [0x55963e3a069a]
    /usr/bin/cataclysm(main+0x1178) [0x55963dca9208]
    /lib64/libc.so.6(__libc_start_main+0xeb) [0x7f4d5aa65ebb]
    /usr/bin/cataclysm(_start+0x2a) [0x55963dd92d7a]

    Attempting to repeat stack trace using debug symbols…
    debug_write_backtrace(std::ostream&)
    …/src/debug.cpp:893
    DebugLog(DebugLevel, DebugClass)
    …/src/debug.cpp:1052
    std::char_traits<char>::length(char const*)
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/char_traits.h:335
    std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/ostream:571
    realDebugmsg(char const*, char const*, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    …/src/debug.cpp:293
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_is_local() const
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/basic_string.h:222
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose()
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/basic_string.h:231
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/basic_string.h:658
    void realDebugmsg<int const&, int const&, int const&>(char const*, char const*, char const*, char const*, int const&, int const&, int const&)
    …/src/debug.h:80
    map::get_submap_at(tripoint const&) const
    …/src/map.cpp:8186
    map::get_submap_at(tripoint const&) const
    …/src/map.cpp:8183
    map::remove_submap_camp(tripoint const&)
    …/src/map.cpp:5580
    basecamp::abandon_camp()
    …/src/faction_camp.cpp:1629
    basecamp::abandon_camp()
    …/src/faction_camp.cpp:1629
    talk_function::abandon_camp(npc&)
    …/src/npctalk_funcs.cpp:375
    __gnu_cxx::__normal_iterator<talk_effect_fun_t const*, std::vector<talk_effect_fun_t, std::allocator<talk_effect_fun_t> > >::operator++()
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/stl_iterator.h:829
    talk_effect_t::apply(dialogue&) const
    …/src/npctalk.cpp:2080
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_is_local() const
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/basic_string.h:222
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose()
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/basic_string.h:231
    std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/basic_string.h:658
    dialogue::opt(dialogue_window&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, talk_topic const&)
    …/src/npctalk.cpp:1546
    bool std::operator==<char, std::char_traits<char>, std::allocator<char> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*)
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/basic_string.h:6177
    avatar::talk_to(std::unique_ptr<talker, std::default_delete<talker> >, bool, bool)
    …/src/npctalk.cpp:751
    std::unique_ptr<talker, std::default_delete<talker> >::~unique_ptr()
    /usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/g++-v9/bits/unique_ptr.h:291
    game::chat()
    …/src/npctalk.cpp:508
    game::handle_action()
    …/src/handle_action.cpp:1992
    game::do_turn()
    …/src/game.cpp:1505
    main
    …/src/main.cpp:709
    __libc_start_main
    ??:?
    _start
    ??:?
Backtrace emission took 9 seconds.
(continued from above) ERROR : src/map.cpp:8186 [submap* map::get_submap_at(const tripoint&) const] Tried to access invalid map position (-5196, -3684, 0)
18:49:50.955 ERROR : src/map.cpp:5581 [void map::remove_submap_camp(const tripoint&)] Tried to remove camp at (-5196,-3684,0) but the submap is not loaded
```

# Steps To Reproduce
Load the game and try to abandon the camp.

# Expected behavior

# Screenshots

# Versions and configuration

- OS: Linux
    - OS Version: <unknown>
- Game Version: 0.E-8128-gc363e3344a-dirty [64-bit]
- Graphics Version: Curses
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Alternative Map Key [alt_map_key],
    No Rail Stations [No_Rail_Stations]
]

# Additional context

[2020.04.12.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5680157/2020.04.12.zip)

### PR DESCRIPTION
**Pull request by Godofbush**
#### Summary

SUMMARY: Bugfixes "fixed abandon faction camp error"

#### Purpose of change

Fixes #45987 

#### Describe the solution

<em>**Why the bug happens**</em>

When you ask npc to start a camp, the camp instance doesn't set its bulletin board position (aka `bb_pos`), until the first time you examine the bulletin board. But abandon action requires `bb_pos`. 

So the reason: empty `bb_pos` causes this error. 

Here's related code snippets:

```c++
// basecamp::abandon_camp()
map &here = get_map();
here.remove_submap_camp( here.getlocal( bb_pos ) );
```

<em>**Solution**</em>

Transform the `omt_pos` to `ms_pos`, use `ms_pos` in place of `bb_pos`;

Related code snippets:

```c++
// basecamp::abandon_camp()
map &here = get_map();
auto sm_pos = omt_to_sm_copy( omt_pos.raw() );
auto ms_pos = sm_to_ms_copy( sm_pos );
here.remove_submap_camp( here.getlocal( ms_pos ) );
```

On `remove_submap_camp` function, use `p` to get `current_submap`, so we can safely use `ms_pos` replace `bb_pos`, they're in the same submap.

```c++
// map::remove_submap_camp( const tripoint &p )
submap *const current_submap = get_submap_at( p );
current_submap->camp.reset();
```

#### Describe alternatives you've considered

Maybe we can set `bb_pos` when start a camp, but I don't know how to get the bulletin board position.

#### Testing

before fix:

1. ask npc to start a camp, examine the bulletin board, ask npc to abandon the camp, it should have no error message.
2. ask npc to start a camp, ask npc to abandon the camp, it should throw the error message.

after fix:

1. ask npc to start a camp, examine the bulletin board, ask npc to abandon the camp, it should have no error message.
2. ask npc to start a camp, ask npc to abandon the camp, it should have no error message.

#### Additional context

None

### CODE REVIEW

File: src/faction_camp.cpp [hunk 1]
```diff
@@ -1618,7 +1618,11 @@ void basecamp::abandon_camp()
     }
     overmap_buffer.remove_camp( *this );
     map &here = get_map();
-    here.remove_submap_camp( here.getlocal( bb_pos ) );
+    auto sm_pos = omt_to_sm_copy( omt_pos.raw() );
+    auto ms_pos = sm_to_ms_copy( sm_pos );
```
Comment (BrettDong) [inline src/faction_camp.cpp:1622]:
```suggestion
    const tripoint &sm_pos = omt_to_sm_copy( omt_pos.raw() );
    const tripoint &ms_pos = sm_to_ms_copy( sm_pos );
```

**Review by BrettDong — CHANGES_REQUESTED** [review]
Usage of `auto` keyword is discouraged here in this repository and there is a `clang-tidy` rule ( see #43370 ) checking for any violations. Have to change `auto` to `tripoint` to pass CI checks.

**Comment by BreadPrince** [comment]
> Usage of `auto` keyword is discouraged here in this repository and there is a `clang-tidy` rule ( see #43370 ) checking for any violations. Have to change `auto` to `tripoint` to pass CI checks.

Thx man, I've learn a lot from you

File: src/faction_camp.cpp [hunk 2]
```diff
@@ -1618,7 +1618,11 @@ void basecamp::abandon_camp()
     }
     overmap_buffer.remove_camp( *this );
     map &here = get_map();
-    here.remove_submap_camp( here.getlocal( bb_pos ) );
+    const tripoint &sm_pos = omt_to_sm_copy( omt_pos.raw() );
+    const tripoint &ms_pos = sm_to_ms_copy( sm_pos );
```
Comment (anothersimulacrum) [inline src/faction_camp.cpp:1622]: These return copies, not references.
```suggestion
    const tripoint sm_pos = omt_to_sm_copy( omt_pos.raw() );
    const tripoint ms_pos = sm_to_ms_copy( sm_pos );
```

Comment (BrettDong) [reply src/faction_camp.cpp:1622]: I was binding the return value from these two functions to local constant references.

**Review by BrettDong — COMMENTED** [review]
I am testing this and notice that after I abandon a camp through NPC chat, when I start a chat again, the options look like as if the camp is still there. Is this a separate bug?
螢幕截圖 2021-01-13 下午8 58 31

**Comment by BreadPrince** [comment]
> I am testing this and notice that after I abandon a camp through NPC chat, when I start a chat again, the options look like as if the camp is still there. Is this a separate bug?
> 螢幕截圖 2021-01-13 下午8 58 31

Yes, this is a seperate bug. Cannot recover an abandoned faction camp.

On `src/faction_camp.cpp:597`, there is a function `talk_function::recover_camp( npc &p )`, but just defined never use.

**Review by BrettDong — APPROVED** [review]

**Pull request merged**